overlord/servicestate/quota_handlers.go

```diff
@@ -576,6 +579,62 @@ type ensureSnapServicesForGroupOptions struct {
 	extraSnaps []string
 }
 
+func snapServiceNames(info *snap.Info) []string {
+	var appNames []string
+	for _, app := range info.Services() {
+		appNames = append(appNames, fmt.Sprintf("%s.%s", info.InstanceName(), app.Name))
+	}
+	return appNames
+}
+
+// affectedSnapServices returns a map of snaps and the services affected
+// by performing changes to a quota group. For groups that contain just snaps, all
+// services are added to the map, for groups that contain specific services, only those
+// services listed in the group are affected, as snaps cannot be in same group as services.
+func affectedSnapServices(st *state.State, grp *quota.Group, opts *ensureSnapServicesForGroupOptions) (map[*snap.Info]*wrappers.SnapServiceOptions, []string, error) {
+
+	snapSvcMap := map[*snap.Info]*wrappers.SnapServiceOptions{}
+	addSnapToMap := func(sn string) (*snap.Info, error) {
+		info, err := snapstate.CurrentInfo(st, sn)
+		if err != nil {
+			return nil, err
+		}
+		if snapSvcMap[info] != nil {
+			return info, nil
+		}
+		opts, err := SnapServiceOptions(st, info, opts.allGrps)
+		if err != nil {
+			return nil, err
+		}
+		snapSvcMap[info] = opts
+		return info, nil
+	}
+
+	// handle extra snaps also passed here, so that they get included
+	// in any case
+	var affectedServices []string
+	for _, sn := range append(grp.Snaps, opts.extraSnaps...) {
+		info, err := addSnapToMap(sn)
+		if err != nil {
+			return nil, nil, err
+		}
+		affectedServices = append(affectedServices, snapServiceNames(info)...)
+	}
+
+	// if the group is a service group, then grp.Snaps is empty and we
+	// need to get the affected snaps from the service names, which are
+	// of format 'snap.service'
+	for _, svc := range grp.Services {
+		parts := strings.Split(svc, ".")
```

Comment:
(nitpick) if we only use the first element, we could use `strings.SplitN(svc, ".", 2)` here too?

Reply:
Ah right, nice suggestion! Makes a lot of sense. Changed this.